scratch_kotlin_lib/src/main/java/dev/sgrem/scratch_kotlin_lib/Week4Homework.kt

```diff
@@ -0,0 +1,160 @@
+        selector: Task.() -> T): List<Task> {
+        return taskList.sortedBy { it.selector() }
+    }
+
+    val tasksSortedByTitle = sortedTasks(taskList) {title}
+    println("Tasks sorted by title:")
+    tasksSortedByTitle.forEach {println(it)}
+    println()
+
+    val tasksSortedByPriority = sortedTasks(taskList) {priority}
+    println("Tasks sorted by priority:")
+    tasksSortedByPriority.forEach {println(it)}
+    println()
+
+    val tasksSortedByCompleted = sortedTasks(taskList) {completed}
+    println("Tasks sorted by completed:")
+    tasksSortedByCompleted.forEach {println(it)}
+    println()
+
+    // 5. Filter Tasks:
+    fun filterTasks(
+        taskList: MutableList<Task>,
+        selector: Task.() -> Boolean
+    ): List<Task>{
+        return taskList.filter { it.selector() }
+    }
+
+    val tasksFilteredByCompleted = filterTasks(taskList) {completed}
+    println("Tasks filtered by completed:")
+    tasksFilteredByCompleted.forEach {println(it)}
+    println()
+
+    // TaskManager object defined at top level
+    TaskManager.addTask("Task #14", TaskStatus.Completed)
+    TaskManager.addTask("Task #10", TaskStatus.InProgress(Priority.LOW))
+    TaskManager.addTask("Task #20")
+    TaskManager.addTask("Task #13", TaskStatus.InProgress(Priority.MEDIUM))
+    TaskManager.addTask("Task #1", TaskStatus.Completed)
+
+    val taskManagerTasksSortedByTitle = TaskManager.sortedTasksByTitle()
+    println("TaskManager Tasks sorted by title:")
+    taskManagerTasksSortedByTitle.forEach {println(it)}
+    println()
+    val taskManagerTasksSortedByStatus = TaskManager.sortedTasksByStatus()
+    println("TaskManager Tasks sorted by status:")
+    taskManagerTasksSortedByStatus.forEach {println(it)}
+    println()
+
+    val TaskManagerTasksFilteredByCompleted = TaskManager.filterTasks()
+    println("TaskManager Tasks filtered by completed:")
+    TaskManagerTasksFilteredByCompleted.forEach {println(it)}
+    println()
+}
+
+// 6. Enum Class
+//    NOTE: An enum class can't be defined in a function as a local class.
+enum class Priority(){
+    LOW,
+    MEDIUM,
+    HIGH
+}
+
+// 7. Sealed Class:
+//    NOTE: A sealed class can't be defined in a function as a local class.
+sealed class TaskStatus {
+    data class InProgress(val priority: Priority) : TaskStatus()
+    data object Completed : TaskStatus()
+}
+
+// 8. Objects
+//    Replaced Task priority and completed with status sealed class TaskStatus
+object TaskManager {
+    data class Task(
+        val title: String,
+        var status: TaskStatus
+    )
+    val taskList: MutableList<Task> = mutableListOf()
+
+    fun addTask(
+        title: String,
+        status: TaskStatus = TaskStatus.InProgress(Priority.HIGH)
+    ){
+        taskList.add(Task(title, status))
+    }
+
+    // Sealed class TaskStatus doesn't have a natural order and
+    // needs a custom comparator to sort based on status
+    val statusComparator = Comparator<Task> { task1, task2 ->
```

Comment:
Great work here on creating a comparator! Note, you could move this up above the addTask() method.  This val along with `taskList` could also be `private` that way you don't expose the underlying property outside of this instance.